server/fleet/hosts.go

```diff
@@ -250,21 +250,142 @@ type MDMHostData struct {
 	Profiles *[]HostMDMAppleProfile `json:"profiles,omitempty" db:"profiles" csv:"-"`
+
+	// MacOSSettings indicates macOS-specific MDM settings for the host, such
+	// as disk encryption status and whether any user action is required to
+	// complete the disk encryption process.
+	//
+	// It is not filled in by all host-returning datastore methods.
+	MacOSSettings *MDMHostMacOSSettings `json:"macos_settings,omitempty" db:"-" csv:"-"`
+}
+
+type DiskEncryptionState string
+
+const (
+	DiskEncryptionApplied             DiskEncryptionState = "applied"
+	DiskEncryptionActionRequired      DiskEncryptionState = "action_required"
+	DiskEncryptionEnforcing           DiskEncryptionState = "enforcing"
+	DiskEncryptionFailed              DiskEncryptionState = "failed"
+	DiskEncryptionRemovingEnforcement DiskEncryptionState = "removing_enforcement"
+)
+
+type ActionRequiredState string
+
+const (
+	ActionRequiredLogOut    ActionRequiredState = "log_out"
+	ActionRequiredRotateKey ActionRequiredState = "rotate_key"
+)
+
+type MDMHostMacOSSettings struct {
+	DiskEncryption DiskEncryptionState `json:"disk_encryption" csv:"-"`
+	ActionRequired ActionRequiredState `json:"action_required" csv:"-"`
+}
+
+// DetermineDiskEncryptionStatus determines the disk encryption status for the
+// host based on the file-vault profile in its list of profiles and whether its
+// disk encryption key is available and decryptable. The file-vault profile
+// identifier is received as argument to avoid a circular dependency.
+func (d *MDMHostData) DetermineDiskEncryptionStatus(profiles []HostMDMAppleProfile, fileVaultIdentifier string) {
+	var settings MDMHostMacOSSettings
+
+	var fvprof *HostMDMAppleProfile
+	for _, p := range profiles {
+		p := p
+		if p.Identifier == fileVaultIdentifier {
+			fvprof = &p
+			break
+		}
+	}
+	if fvprof != nil {
+		switch fvprof.OperationType {
+		case MDMAppleOperationTypeInstall:
+			switch {
+			case fvprof.Status != nil && *fvprof.Status == MDMAppleDeliveryApplied:
+				if d.rawDecryptable != nil && *d.rawDecryptable == 1 {
+					//  if a FileVault profile has been successfully installed on the host
+					//  AND we have fetched and are able to decrypt the key
+					settings.DiskEncryption = DiskEncryptionApplied
+				} else if d.rawDecryptable != nil {
+					// if a FileVault profile has been successfully installed on the host
+					// but either we didn't get an encryption key or we're not able to
+					// decrypt the key we've got
+					settings.DiskEncryption = DiskEncryptionActionRequired
+					if *d.rawDecryptable == 0 {
+						settings.ActionRequired = ActionRequiredRotateKey
+					} else {
+						settings.ActionRequired = ActionRequiredLogOut
+					}
+				} else {
+					// if [a FileVault profile is pending to be installed or] the
+					// matching row in host_disk_encryption_keys has a field decryptable
+					// = NULL
+					settings.DiskEncryption = DiskEncryptionEnforcing
+				}
+
```

Comment:
The codecov warnings are quite annoying, that's because this is covered in the tests outside this package so it doesn't see it (all states are tested in `server/service/hosts_test.go` below).

Reply:
yes, they make code review difficult too because they're often in the way.